### PR TITLE
Call procedures from within Cypher queries

### PR DESCRIPTION
**Pull request by boggle**
This change allows queries like

```
 MATCH (e:Episode)
 CALL my.awesome.proc("energize", e.id) YIELD maxWarp AS factor, tribbles AS pets
 RETURN factor, count(pets)
```

This works now (assuming the build is green) but is not yet complete:
- The CLG still needs to finalize syntax
- There are some TODOs left and tests missing

The architecturally most interesting change is moving the first round of semantic checking over to plan building. This became necessary since type checking now depends on information obtained from the planning context - namely the procedure's signature. Even though today that's static it's quite plausible to assume it will by dynamic in the mid term future hence the move seemed like right thing to do.

The commits meander a bit as it took me a while to figure this out and solidify the model used.

Also noteworthy, this change preserves the fast code path developed for just calling procedures and only goes through the old run time when Cypher and procedures are mixed. This change does not add procedure calling to birk, this can happen later.

### CODE REVIEW

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/UnwindPipe.scala [hunk 1]
```diff
@@ -82,3 +82,6 @@ case class UnwindPipe(source: Pipe, collection: Expression, variable: String)
     }
   }
 }
+
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/UnwindPipe.scala:85]: ?

Comment (boggle) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/UnwindPipe.scala:85]: No idea why :)

**Comment by kvegter** [comment]
This would be very useful functionality. This is the 'missing link'. When I deployed my first procedure on (m05 (nightly build)) I directly tried to use the call statement within a cypher statement...

**Comment by davidegrohmann** [comment]
It would be nice to add some acceptance tests in `EagerizationAcceptanceTest` to ensure that everything is working as expected.

**Comment by boggle** [comment]
@davidegrohmann I'll add that later tonight

**Comment by boggle** [comment]
@davidegrohmann added test per you request, though something is wrong with TC not picking this up... re-opened as: https://github.com/neo4j/neo4j/pull/6659

**Pull request closed without merging**